### ISSUE TITLE
Wishlist: `igraph_is_complete()` decides whether a graph is complete

### ISSUE THREAD

**Issue by aagon** [igraph/igraph#2490]
I was surprised to see that there was no function to decide whether a graph is
complete or not. Of course, one might decide to compute the largest clique of
the graph and compare it to the graph itself, but that would be very costly.

The context: I am generating a lot of random geometric graphs for my
experiments, and I simply wanted to discard the complete ones. I gave myself a
very simple implementation that compares the amount of vertices and edges and
concludes (I'm only dealing with simple graphs).

I thought I would contribute an implementation of such a function, but I would
like to see whether you're interested first, and I had a couple of questions.

The heuristic would be the following:

- If the amount of edges is low enough that the graph cannot possibly be complete,
  we return early (O(1)).

- Then, we test if the graph is simple, with `igraph_is_simple()` O(e + v) in
  the worst case, O(1) if cache hit. If the graph is simple, we can compare and
  conclude.
  
  (Side question here: why do you cache the result of `igraph_is_simple()` only
  if it is positive ?)

- If the graph is not simple, we count the loops with function
  `igraph_is_loop()` (O(e)).
  
- We test again whether the amount of edges (minus the loops) might not be low
  enough for us to conclude directly (O(1)).

- If it is not the case, we need to test whether there are multiple edges with
  `igraph_has_multiple()`. This should be O(1), since we should have cached the
  result from `igraph_is_simple()`.
  
- If there are none, we can conclude by simple comparison.

- If there are some, we use `igraph_is_multiple()` to count the multiple edges.
  (O(e*d)) with d the average degree of the graph.
  
- We can now conclude by comparing and accounting for both loops and multiple
  edges.

This heuristic is maybe a little complicated, but ensures that we conclude as
soon as we can, and tries to shove the expensive operations at the end.

Also, the memory consumption is minimal. We only need one vector of size e, to
store the result of functions `igraph_is_loop()` and `igraph_is_multiple()`.

I also had in mind another, simpler heuristic that would duplicate the graph,
and then simplify the duplicate with `igraph_simplify()`, but this would require
something like O(e + v) memory for the new graph.

I'll let you decide whether you prefer to optimise for memory or simplicity.

I wanted to ask you a couple of questions before proceeding :

- Are you interested at all by such a function ?

- In what file would you want me to put it ?

- A complete undirected graph is a graph such that there is an undirected edge
  between every couple of vertices, but a complete directed graph can
  potentially admit different definitions : Is it such that there are 2 directed
  edges, one in each direction, between each couple of vertices ? Or only 1
  directed edge would be enough ("weakly" complete) ?

- There was a discussion regarding null graphs and connectivity, and it was
  decided to consider null graphs not connected. Do I consider null graphs
  complete or not ?

- Would you be interested in a PR to improve `igraph_is_simple()`, in order to
  make it cache whether there are loops or multiple edges regardless of its
  result ?

Thank you in advance for your patience !

Best,

Aymeric

**Comment by szhorvat** [comment]
I already had a feature request open for this, see #2380. I would like to make the interface general, and be able to test if groups of vertices form cliques.  If there's an opportunity to perform the check faster for the entire graph, we should take it.

You are welcome to work on this. I suggest keeping it simple first, and adding a comprehensive test suite to make sure that all edge cases are covered. Then we can optimize it later.

As for why a negative result from `igraph_is_simple()` is not cached: There is no corresponding cache entry. We have separate `HAS_LOOP` and `HAS_MULTIPLE` entries.  If all we know is that the graph is not simple, we're still lacking the information about why: is it because it had loop? or because it had multi-edges?  Of course this could be improved by chanhing `igraph_is_simple()` to record what it found. We are being conservative with setting the cache, and improving on this very slowly, as cache bugs can have very nasty consequences.

----

Let's set down the definition of a clique / complete graph to make sure we're on the same page. We require that all pairs of vertices are adjacent, but we don't care by how many edges are present between them (it can be one or more). We ignore self-loops.

-----

The procedure:

> * If the amount of edges is low enough that the graph cannot possibly be complete,
>       we return early (O(1)).

Sounds great.

Also, as you said, if we know that the graph is simple from the cache, we only need to count edges and we're done.

For the rest of the procedure, I don't think we can avoid $O(|E|)$, which is unfortunately, as $|E| \sim |V|^2$ for graphs on which this function is likely to be called.

If we need to iterate through all edges anyway, it's best to do it in a single loop instead of calling multiple subroutines, each of which performs a similar iteration. The _simplest_ implementation here is to call `igraph_i_neighbors()` on all vertices with `IGRAPH_NO_LOOPS` and `IGRAPH_NO_MULTIPLE`, then simply check the length of the neighbour vector. An improvement over this would be to iterate through the neighbour vectors and check that there is a connection to all other vertices. Note that the neighbour vector is guaranteed to be sorted, so this will be simple to do.

Overall this will be $O(|E| + |V|)$ at worst.

Let me know what you think, and whether you'd like to work on this.

----

> * Are you interested at all by such a function ?

Absolutely.

> * In what file would you want me to put it ?

A new file, maybe `properties/complete.c`.

> * A complete undirected graph is a graph such that there is an undirected edge
>   between every couple of vertices, but a complete directed graph can
>   potentially admit different definitions : Is it such that there are 2 directed
>   edges, one in each direction, between each couple of vertices ? Or only 1
>   directed edge would be enough ("weakly" complete) ?

In the longer terms we could add a parameter `igraph_bool_t directed`. If `false`, edge directions are ignored in directed graphs. If `true`, the graph is complete only if connections are present in both directions.

But this brings complications and increases the chance of mistakes, so I would do this only at a later stage. Note that a directed graph will be "simple" with edge directions ignored only if `igraph_has_mutual()` is false. `igraph_is_simple()` does not suffice here.
 
> * There was a discussion regarding null graphs and connectivity, and it was
>   decided to consider null graphs not connected. Do I consider null graphs
>   complete or not ?

At the moment I cannot think of a good reason not to. It fits the basic definition of "all pairs are connected". 

I would choose considering the null graph complete.

> 
> * Would you be interested in a PR to improve `igraph_is_simple()`, in order to
>   make it cache whether there are loops or multiple edges regardless of its
>   result ?

I will do this.

**Comment by aagon** [comment]
First of all, thank you for your quick and detailed answer (and sorry about my late answer).

> As for why a negative result from igraph_is_simple() is not cached: There is no corresponding cache entry. We have separate HAS_LOOP and HAS_MULTIPLE entries. If all we know is that the graph is not simple, we're still lacking the information about why: is it because it had loop? or because it had multi-edges? Of course this could be improved by chanhing igraph_is_simple() to record what it found. We are being conservative with setting the cache, and improving on this very slowly, as cache bugs can have very nasty consequences.

I should have been clearer. When I said "negative result from `igraph_is_simple()`", I indeed meant the fact that it could either have loops or multiple edges, for which there are cache entries.

> Note that the caching can still only be partial. As soon as we found a loop, we know it's not a simple graph, and we can return. We do not need to keep going and look for multi-edges. So we won't always get information about both properties.

Exactly. But I would argue that this is better than caching nothing.

> I will improve on this.

This is what I had in mind, what do you think :

```diff
diff --git a/src/properties/multiplicity.c b/src/properties/multiplicity.c
index a7e4ad9c1..b9cc3f9c3 100644
--- a/src/properties/multiplicity.c
+++ b/src/properties/multiplicity.c
@@ -74,10 +74,18 @@ igraph_error_t igraph_is_simple(const igraph_t *graph, igraph_bool_t *res) {
             n = igraph_vector_int_size(&neis);
             for (j = 0; j < n; j++) {
                 if (VECTOR(neis)[j] == i) {
-                    found = true; break;
+                    found = true;
+                    igraph_i_property_cache_set_bool(graph,
+                                                     IGRAPH_PROP_HAS_LOOP,
+                                                     true);
+                    break;
                 }
                 if (j > 0 && VECTOR(neis)[j - 1] == VECTOR(neis)[j]) {
-                    found = true; break;
+                    found = true;
+                    igraph_i_property_cache_set_bool(graph,
+                                                     IGRAPH_PROP_HAS_MULTI,
+                                                     true);
+                    break;
                 }
             }
         }
```


> You are welcome to work on this.

To be clear, do you mean the function `igraph_is_clique()` or its specific case
`igraph_is_complete()` (or both) ?

> The simplest implementation here is to call igraph_i_neighbors() on all
> vertices with IGRAPH_NO_LOOPS and IGRAPH_NO_MULTIPLE, then simply check the
> length of the neighbour vector.

This is O(v * d), the cost of `igraph_i_neighbors()` on all vertices.

> An improvement over this would be to iterate through the neighbour vectors and
> check that there is a connection to all other vertices.

I assume that by "neighbour vectors", you mean the set of edges of the graph.
Indeed, iterating over those would allow to :
- treat each edge only once
- discarding loops and multiple edges as we go
- concluding early if the current subset of edges departing from a vertex is too
  small

So O(e) at worst.

However, to do this, we would need the set of edges to be sorted by first
adjacent vertex, then second adjacent vertex, and to always follow the
convention "smallest VID -- largest VID". Is this what you mean by the
following :

> Note that the neighbour vector is guaranteed to be sorted, so this will be
> simple to do.

> But this brings complications and increases the chance of mistakes, so I would
> do this only at a later stage.

Understood, I'll error with IGRAPH_UNIMPLEMENTED in this case.

> Let me know what you think, and whether you'd like to work on this.

I'd like to work on this, I'll get to it.

**Comment by szhorvat** [comment]
> I'd like to work on this, I'll get to it.

Great! Let me just give quick response on this so you can get started. I'll try to get back later with a more detailed response.

Regarding the cache, I have a project going to improve the cache significantly. This is one reason I'd like to postpone this a little bit. More on this later.l



> To be clear, do you mean the function `igraph_is_clique()` or its specific case
> `igraph_is_complete()` (or both) ?

You are welcome to work on both. Let's implement the functionality and we can think about how to package it later. We'll also keep the functions experimental for the moment, which means that we reserve the right to make changes. You were interested in checking if an entire graph is complete, so I assume you'd like to start with that?



> However, to do this, we would need the set of edges to be sorted by first
> adjacent vertex, then second adjacent vertex, and to always follow the
> convention "smallest VID -- largest VID". Is this what you mean by the
> following :

Yes, `igraph_neighbors()` (and the `_i_` version) guarantee to return sorted neighbour lists. The internal graph representation is such that it makes it efficient to retrieve neighbour lists.

It's clear that you already have experience with igraph, but I'll link this anyway: https://github.com/igraph/igraph/wiki/Quickstart-for-new-contributors

**Comment by aagon** [comment]
> Great! Let me just give quick response on this so you can get started.

I have a minimal working implementation on `igraph_is_complete`, see #2510. I'll get started on `igraph_is_clique` after we're done with this.

> The simplest implementation here is to call igraph_i_neighbors() on all vertices with IGRAPH_NO_LOOPS and IGRAPH_NO_MULTIPLE, then simply check the length of the neighbour vector.

I have done this. I'm afraid I do not get how your second proposal is an improvement over this.

**Comment by szhorvat** [comment]
Implemented in #2510, thanks @aagon !
